### PR TITLE
Make lang items private

### LINKED ISSUE #72240: Make lang items private
#72170 and #72216 put some effort in to try and prevent some potential ICEs that can occur in `#![no_core]` due to how we currently reference lang items.

@oli-obk pointed out:

> Oh, that's a very good idea. I wonder if we can make `lang_items` private at some point and just have sensible wrappers like `require_lang_item`.

I'm opening this issue to start a discussion about how we could take this forward.

### PR DESCRIPTION
**Pull request by cofibrant**
Implements https://github.com/rust-lang/compiler-team/issues/301

Design is largely as discussed, but I also made an effort to remove `require_lang_item`. Instead, `LangItemRecord` has methods `require` and `require_with` which take a reference to a `MissingLangItemHandler` (implemented by `TyCtxt`). In my opinion this makes the interface a lot cleaner while also reducing the number of imports (it is no longer necessary to import `lang_items` or specific items in order to require them).

This change also made it possible to unify the `item` and `missing` fields of the lang items table. We no longer require the hack in the weak lang items pass for marking `eh_personality` as missing as all lang items are now assumed missing until found. To make this work, `alloc_error_handler` is explicitly whitelisted in crates which do not require an allocator.

This is likely to conflict with a lot of open PRs as it touches a lot of code, but my biggest worry is #72559 (which will need to be re-implemented to use the new API although I can't imagine it will be too difficult).

Resolves #72240

r? @oli-obk 

cc @ecstatic-morse + @Aaron1011

### CODE REVIEW

**Comment by oli-obk** [comment]
What do you think will be easier to do? Merge this PR first or merge #72240 first?

**Comment by doctorn** [comment]
> What do you think will be easier to do? Merge this PR first or merge #72240 first?

Probably this as I think patching up #72240 will just be a case of replacing some instances of `Some` with `LangItemRecord::Present`, but I haven't had a chance to read it properly yet

**Comment by doctorn** [comment]
> What do you think will be easier to do? Merge this PR first or merge #72559 first?

I'll give it a proper read this afternoon but my gut feeling is probably this

**Comment by oli-obk** [comment]
I'll review the other PR then and carry it over the finish line

File: src/librustc_hir/lang_items.rs [hunk 1]
```diff
@@ -109,6 +187,12 @@ impl<CTX> HashStable<CTX> for LangItem {
     }
 }
 
+impl<CTX> HashStable<CTX> for LangItemRecord {
+    fn hash_stable(&self, _: &mut CTX, hasher: &mut StableHasher) {
+        ::std::hash::Hash::hash(self, hasher);
+    }
```
Comment (lcnr) [inline src/librustc_hir/lang_items.rs:193]: Is this implementation correct? `LangItemRecord` can contain a `DefId` for which `hash` is not stable.

Comment (ecstatic-morse) [reply src/librustc_hir/lang_items.rs:193]: Good catch.

**Comment by JohnCSimon** [comment]
Ping from triage:
@doctorn Can you please address the merge conflicts and build failures.

**Comment by Muirrum** [comment]
@doctorn This is another ping from Triage.

**Comment by Dylan-DPC** [comment]
@doctorn hey :) closing this due to inactivity. If you wish when you have the time to do so, you can submit a new PR with these changes. Thank you for taking the time to contribute :)

**Pull request closed without merging**